### PR TITLE
Improve Error Handling to Prevent Exit, Crashes and Memory Leaks

### PR DESCRIPTION
**Pull request by songlingatpan**
This PR focuses on improving the robustness and stability of the code by enhancing error handling. The main objectives of this update include:

- Preventing the program from exiting or crashing in case of errors
- Preventing memory leaks by checking for NULL pointers after memory allocations and ensuring proper memory cleanup in failure cases.
- Adding safeguards for potential memory allocation issues, ensuring the program doesn't crash or exit unexpectedly.

Note: This PR does not introduce any new functionality but aims to stabilize the existing code and make it more resilient to errors. Future test cases will be added to simulate scenarios such as out-of-memory conditions and ensure these improvements work as intended.

### CODE REVIEW

**Comment by songlingatpan** [comment]
This PR depends on #1823. Since the changes from #1823 have not been merged yet, I have cherry-picked those commits into this PR. Once #1823 is merged, I will rebase this PR on top of the main branch to ensure a clean merge.

**Comment by baentsch** [comment]
@songlingatpan Thanks for the various enhancements. I now triggered a more complete CI run. Do you expect to do further commits or is this now indeed Ready for Review from your perspective?

File: src/common/common.c [hunk 1]
```diff
@@ -298,16 +300,37 @@ void *OQS_MEM_checked_aligned_alloc(size_t alignment, size_t size) {
...
+	if (!size) {
+		return NULL;
+	}
+	const size_t offset = alignment - 1 + sizeof(uint8_t);
+	uint8_t *buffer = OPENSSL_malloc(size + offset);
+	if (!buffer) {
+		return NULL;
+	}
+	uint8_t *ptr = (uint8_t *)(((uintptr_t)(buffer) + offset) & ~(alignment - 1));
+	ptrdiff_t diff = ptr - buffer;
```
Comment (baentsch) [inline src/common/common.c:323]: Looks like quite a few compilers don't "appreciate" this type. See CI results...

**Comment by songlingatpan** [comment]
> @songlingatpan Thanks for the various enhancements. I now triggered a more complete CI run. Do you expect to do further commits or is this now indeed Ready for Review from your perspective?

I will post more commits. Thanks

**Comment by songlingatpan** [comment]
@dstebila @baentsch Could you please trigger the complete CI run again? Thanks

**Comment by songlingatpan** [comment]
@baentsch CI run again? Thanks

**Comment by baentsch** [comment]
> I will post more commits.

Then please move this PR status to "Draft".

> CI run again? 

Done. Other failures. Could you please try to [get CI to pass locally](https://github.com/open-quantum-safe/liboqs/blob/main/CONTRIBUTING.md#running-ci-locally) first so that I'm not required to manually trigger re-runs any more @songlingatpan ? If all seems to be OK, please revert to Ready for Review stage. Thanks.

**Comment by songlingatpan** [comment]
@dstebila @baentsch 
This PR is ready for review and merge. Thanks

**Comment by baentsch** [comment]
Also closing due to inactivity. If you want to re-start, please re-open or create a new PR on current "main" to avoid rebasing, @songlingatpan .

**Pull request closed without merging**